builtin/joined_dataset.cc

```diff
@@ -181,115 +159,8 @@ struct JoinedDataset::Itl
-                }
-                break;
-            }
-            case AnnotatedJoinCondition::EQUIJOIN: {
-                // Join on f(leftrow) = f(rightrow)
-                break;
-            }
-            default:
-                throw HttpReturnException(400, "Join expression requires an equality operator; needs to be in the form f(left) = f(right)",
-                                          "joinOn", joinConfig.on,
-                                          "condition", condition);
-            }
-
-            // Finally, perform the join
-            // We keep a list of the row hashes of those that join up
-            auto it1 = leftRows.begin(), end1 = leftRows.end();
-            auto it2 = rightRows.begin(), end2 = rightRows.end();
-
-            while (it1 != end1 && it2 != end2) {
-                // TODO: there could be multiple values...
-                const ExpressionValue & val1 = std::get<0>(*it1);
-                const ExpressionValue & val2 = std::get<0>(*it2);
-
-                if (debug)
-                    cerr << "joining " << jsonEncodeStr(val1) << " and " << jsonEncodeStr(val2) << endl;
-
-                if (val1 < val2) {
-                    ++it1;
-                }
-                else if (val2 < val1) {
-                    ++it2;
-                }
-                else {
-                    ExcAssertEqual(val1, val2);
-
-                    // We got a match on the join condition.  So now
-                    // we take the cross product of the matching rows.
-
-                    // First figure out how many of each are there?
-                    auto erng1 = it1;  ++erng1;
-                    while (erng1 < end1 && std::get<0>(*erng1) == val1)
-                        ++erng1;
-
-                    auto erng2 = it2;  ++erng2;
-                    while (erng2 < end2 && std::get<0>(*erng2) == val2)
-                        ++erng2;
-
-                    // Now the cross product
-                    if (!val1.empty()) {
-                        for (auto it1a = it1; it1a < erng1;  ++it1a) {
-                            for (auto it2a = it2; it2a < erng2;  ++it2a) {
-                                const RowName & leftName = std::get<1>(*it1a);
-                                const RowName & rightName = std::get<1>(*it2a);
-                                const RowHash & leftHash = std::get<2>(*it1a);
-                                const RowHash & rightHash = std::get<2>(*it2a);
-
-                                if (debug)
-                                    cerr << "rows " << leftName << " and " << rightName << " join on value " << val1 << endl;
-                                
-                                gotRow(leftName, leftHash, rightName, rightHash);
-                            }
-                        }
-                    }
+            makeJoinFast(condition, context, left, right, joinConfig.qualification);            
```

Comment:
Consider renaming the function to mention that this is a join with a constant where.  The fact that this is handling the easier and faster case could be put as a comment.